File: JET/modul_2_oop/L02Datenkapselung/Exercise2/Hamster.cs
namespace Hamster;

public class Hamster
{
    // Felder
    private static string _hungryRepresentation = "😡";
    private static string _fedRepresentation = "🐹";

    // Eigenschaften (Properties)
    public (int x, int y) Position { get; set; }
    public string Representation { get; private set; }

    public bool IsHungry { get; private set; }

    // Beziehungen
    private Plane plane;
    private List<Seed> mouth = new List<Seed>();

    // Konstruktor
    public Hamster(Plane plane)
    {
        IsHungry = false;
        Representation = _fedRepresentation;
        this.plane = plane;

        PositionAndManageHamster();
    }

    // Methoden
    private void PositionAndManageHamster()
    {
        var random = new Random();
        bool done;
        int x, y;

        do
        {
            x = random.Next(plane.Size);
            y = random.Next(plane.Size);

            done = plane.AssignInitialPosition(this, (x, y));
        } while (!done);

        Position = (x, y);
    }

    public void Move()
    {
        var random = new Random();
        int index = random.Next(Enum.GetValues<Direction>().Length);
        var direction = Enum.GetValues<Direction>()[index];

        plane.Position(this, direction);
    }

    public void NutritionBehaviour()
    {
        var random = new Random();

        // Zufällig hungrig werden
        if (random.NextDouble() < 0.1)
        {
            IsHungry = true;
            Representation = _hungryRepresentation;
        }

        if (plane.ContainsSeed(Position))
        {
            if (IsHungry)
            {
                EatSeedFromTile();
            }
            else
            {
                PutInMouth();
            }
        }
        else
        {
            if (IsHungry && mouth.Any())
            {
                EatSeedFromMouth();
            }
        }
    }

    private void EatSeedFromMouth()
    {
        Eat();
        mouth.RemoveAt(0);
    }

    public void EatSeedFromTile()
    {
        Eat();
        plane.HamsterIsEatingSeeds(this);
    }

    private void Eat()
    {
        IsHungry = false;
        Representation = _fedRepresentation;
    }

    public void PutInMouth()
    {
        var samen = plane.GetSeedOn(Position);
        mouth.Add(samen);
        plane.HamsterIsStoringSeeds(this);
    }

    public override string ToString()
    {
        return Representation;
    }
}

File: BITEC/modul_2_object_orientation/src/lerneinheiten/L02KlassenUndMethoden/live/hamster/Hamster.java
package lerneinheiten.L02KlassenUndMethoden.live.hamster;

import java.util.ArrayList;
import java.util.Random;

public class Hamster {
    // Felder
    String feldZuMerken = Plane.getEarthRepresentation();
    int xPosition;
    int yPosition;
    String _representation;
    final static String _hungryRepresentation = "🤬";
    final static String _fedRepresentation = "🐹";
    boolean isHungry;
    ArrayList<Seedling> _mouth = new ArrayList<>();

    // Hat-Beziehungen
    Plane _plane;
    // -----------------------------
    // Methoden
    void nahrungsVerhalten() {
        Random random = new Random();

        if (random.nextDouble() < 0.1) {
            isHungry = true;
            _representation = _hungryRepresentation;
        }

        boolean tileHasSeedling = _plane.tileTakenBySeedling(xPosition, yPosition);

        if (tileHasSeedling) {
            if (isHungry) {
                _representation = _fedRepresentation;
                isHungry = false;

                Seedling seedlingAmGleichenFeld = _plane.getSeedlingOnPosition(xPosition, yPosition);
                _plane.getSeedlings().remove(seedlingAmGleichenFeld);
                // hamster speicher seedling im mund...
            }
        }
    }

    void bewegen() {
        Random random = new Random();
        int index = random.nextInt(0, Richtung.values().length);
//        Richtung richtung = Richtung.OBEN;
        Richtung richtung = Richtung.values()[index];

        _plane.bewegeHamster(this, richtung);
    }

    // Konstruktor
    Hamster(Plane plane) {
        boolean done;
        _plane = plane;
        _representation = _hungryRepresentation;
        isHungry = true;
        int xWunsch;
        int yWunsch;

        // zufällig den Hamster positionieren.
        do {
            Random random = new Random();
            xWunsch = random.nextInt(0, _plane.getSize());
            yWunsch = random.nextInt(0, _plane.getSize());

            done = _plane.assignInitialPosition(this, xWunsch, yWunsch);
        } while (!done);

        xPosition = xWunsch;
        yPosition = yWunsch;
    }
}
